indexing-service/src/main/java/io/druid/indexing/common/task/ArchiveTask.java

```diff
@@ -91,12 +98,15 @@ public TaskStatus run(TaskToolbox toolbox) throws Exception
       log.info("OK to archive segment: %s", unusedSegment.getIdentifier());
     }
 
-    // Move segments
-    for (DataSegment segment : unusedSegments) {
-      final DataSegment archivedSegment = toolbox.getDataSegmentArchiver().archive(segment);
-      toolbox.getTaskActionClient().submit(new SegmentMetadataUpdateAction(ImmutableSet.of(archivedSegment)));
+    if (toolbox.getTaskActionClient().submit(new LockUpgradeAction(getInterval()))) {
+      // Move segments
+      for (DataSegment segment : unusedSegments) {
+        final DataSegment archivedSegment = toolbox.getDataSegmentArchiver().archive(segment);
```

Comment:
And for the comment about "realtime task waiting for Hadoop index task acting in critical section". This is not a problem right as Hadoop Index task just publishes segment in MetadataStorage in critical section which should be very quick and RealtimeIndex Task always retries while trying to acquire a lock